oasislmf/pytools/gul/manager.py

```diff
@@ -235,190 +205,126 @@ def run(run_dir, ignore_file_type, sample_size, loss_threshold, alloc_rule, debu
         alloc_rule (int): back-allocation rule.
         rndms (numpy.array[float64]): 2d array of shape (number of seeds, sample_size) storing the random values
           drawn for each seed.
-        rndms_idx (Dict[int]): dic storing the map between the `seed` value and the row in the `rndms` array
-          containing the drawn random samples.
         debug (bool): if True, for each random sample, print to the output stream the random value
           instead of the loss.
         buff_size (int): size in bytes of the output buffer.
         int32_mv (numpy.ndarray): int32 view of the memoryview where the output is buffered.
         cursor (int): index of int32_mv where to start writing.
-        cursor_bytes (int): number of bytes written in int32_mv.
-
+        
     Returns:
         int, int, int: updated value of cursor, updated value of cursor_bytes, last last_processed_coverage_ids_idx
     """
-    for coverage_id in coverage_ids[last_processed_coverage_ids_idx:]:
-        tiv = coverages[coverage_id - 1]  # coverages are indexed from 1
-        Nitems = len(items_data_by_coverage_id[coverage_id])
-        exposureValue = tiv / Nitems
-
-        Nitem_ids = len(item_ids_by_coverage_id[coverage_id])
+    max_size_per_item = (sample_size + NUM_IDX + 1) * gulSampleslevelRec_size + 2 * gulSampleslevelHeader_size
+    for coverage_i in range(last_processed_coverage_ids_idx, coverage_ids.shape[0]):
+        coverage = coverages[coverage_ids[coverage_i]]
+        tiv = coverage['tiv']  # coverages are indexed from 1
+        Nitem_ids = coverage['cur_items']
+        exposureValue = tiv / Nitem_ids
 
         # estimate max number of bytes are needed to output this coverage
         # conservatively assume all random samples are printed (losses>loss_threshold)
         # number of records of type gulSampleslevelRec_size is sample_size + 5 (negative sidx) + 1 (terminator line)
-        est_cursor_bytes = Nitem_ids * (
-            (sample_size + NUM_IDX + 1) * gulSampleslevelRec_size + 2 * gulSampleslevelHeader_size
-        )
+        est_cursor_bytes = Nitem_ids * max_size_per_item
 
         # return before processing this coverage if bytes to be written in mv exceed `buff_size`
-        if cursor_bytes + est_cursor_bytes > buff_size:
-            return cursor, cursor_bytes, last_processed_coverage_ids_idx
+        if cursor * int32_mv.itemsize + est_cursor_bytes > buff_size:
+            return cursor, cursor * int32_mv.itemsize, last_processed_coverage_ids_idx
 
-        # sort item_ids (need to convert to np.array beforehand)
-        item_ids = np.empty(Nitem_ids, dtype=ITEM_ID_TYPE)
-        for j in range(Nitem_ids):
-            item_ids[j] = item_ids_by_coverage_id[coverage_id][j]
+        items = items_data[coverage['start_items']: coverage['start_items'] + coverage['cur_items']]
 
-        item_ids_argsorted = np.argsort(item_ids)
-        item_ids_sorted = item_ids[item_ids_argsorted]
```

Comment:
I've just found that here actually I was sorting the items within each coverage by `item_id`.